pam/integration-tests/gdm-integration_test.go

```diff
@@ -0,0 +1,471 @@
+		cmd.Args = append(cmd.Args, "-asan")
+	}
+
+	libPath := filepath.Join(t.TempDir(), "libpam_authd.so")
+	t.Logf("Compiling PAM library at %s", libPath)
+
+	cmd.Args = append(cmd.Args, "-tags=pam_debug,pam_gdm_debug", "-o", libPath)
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err, string(out))
+
+	return libPath
+}
+
+func buildPAMWrapperModule(t *testing.T) string {
+	t.Helper()
+
+	compiler := os.Getenv("CC")
+	if compiler == "" {
+		compiler = "cc"
+	}
+
+	//nolint:gosec // G204 it's a test so we should allow using any compiler safely.
+	cmd := exec.Command(compiler)
+	soname := "pam_authd_loader"
+	libPath := filepath.Join(t.TempDir(), soname+".so")
+
+	require.NoError(t, os.MkdirAll(filepath.Dir(libPath), 0700))
+	t.Logf("Compiling PAM Wrapper library at %s", libPath)
+	cmd.Args = append(cmd.Args, []string{
+		"-o", libPath,
+		"../go-loader/module.c",
+		"-g3",
+		"-O0",
+	}...)
+
+	if modulesPath := os.Getenv("AUTHD_PAM_MODULES_PATH"); modulesPath != "" {
+		cmd.Args = append(cmd.Args, fmt.Sprintf("-DAUTHD_PAM_MODULES_PATH=%q",
+			os.Getenv("AUTHD_PAM_MODULES_PATH")))
+	}
+	if pam_test.IsAddressSanitizerActive() {
+		cmd.Args = append(cmd.Args, "-fsanitize=address,undefined")
+	}
+	if cflags := os.Getenv("CFLAGS"); cflags != "" && os.Getenv("DEB_BUILD_ARCH") == "" {
+		cmd.Args = append(cmd.Args, strings.Split(cflags, " ")...)
+	}
+
+	cmd.Args = append(cmd.Args, []string{
+		"-Wl,--as-needed",
+		"-Wl,--allow-shlib-undefined",
+		"-shared",
+		"-fPIC",
+		"-Wl,--unresolved-symbols=report-all",
+		"-Wl,-soname," + soname + "",
+		"-lpam",
+	}...)
+	if ldflags := os.Getenv("LDFLAGS"); ldflags != "" && os.Getenv("DEB_BUILD_ARCH") == "" {
+		cmd.Args = append(cmd.Args, strings.Split(ldflags, " ")...)
+	}
+
+	if testutils.CoverDir() != "" {
+		cmd.Args = append(cmd.Args, "--coverage")
+		cmd.Args = append(cmd.Args, "-fprofile-abs-path")
+
+		t.Cleanup(func() {
+			t.Log("Running gcov...")
+			gcov := exec.Command("gcov")
+			gcov.Args = append(gcov.Args,
+				"-pb", "-o", filepath.Dir(libPath),
+				soname+".so-module.gcno")
+			gcov.Dir = testutils.CoverDir()
+			out, err := gcov.CombinedOutput()
+			require.NoError(t, err, string(out))
+			t.Log(string(out))
+		})
+	}
+
+	t.Logf("Running compiler command: %s %s", cmd.Path, strings.Join(cmd.Args[1:], " "))
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err, string(out))
+
+	return libPath
+}
+
+func createServiceFile(t *testing.T, name string, libPath string, args []string, ignoreError string) string {
+	t.Helper()
+
+	serviceFile := filepath.Join(t.TempDir(), name)
+	t.Logf("Creating service file at %s", serviceFile)
+
+	err := os.WriteFile(serviceFile,
+		[]byte(fmt.Sprintf(`auth [success=done ignore=ignore default=die] %[1]s %[2]s
+auth requisite pam_debug.so auth=%[3]s
+account [success=done ignore=ignore default=die] %[1]s %[2]s
+account requisite pam_debug.so acct=%[3]s`, libPath, strings.Join(args, " "), ignoreError)),
+		0600)
+	require.NoError(t, err)
```

Comment:
Ditto I would add more information on the error like "Setup: could not create service file"